### PR TITLE
Bug 1768782: fix(pipeline-run-row) | Duration field in PipelineRun Row List

### PR DESCRIPTION
**Pull request by abhinandan13jan**
Addresses Bug - https://jira.coreos.com/browse/ODC-2181

Adds unit test for the duration utility.

Screencast from 11-08-2019 02_46_00 PM

Also fixes the issue mentioned in Bug - https://jira.coreos.com/browse/ODC-2120

Screenshot from 2019-11-08 14-50-04

### CODE REVIEW

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 1]
```diff
@@ -290,3 +290,21 @@ export const constructCurrentPipeline = (
     status,
   };
 };
+
+export const pipelineRunDuration = (run: PipelineRun): string => {
+  const startTime = _.get(run, ['status', 'startTime'], null);
+  const completionTime = _.get(run, ['status', 'completionTime'], null);
+
+  // Duration cannot be computed if start time is missing or a completed/failed pipeline has no end time
+  if (!startTime || (!completionTime && pipelineRunFilterReducer(run) !== 'Running')) {
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts]: Recommend we start using `pipelineRunStatus` which is what `pipelineRunFilterReducer` calls when we want just the status value (and not a pretty display value). Obviously it doesn't matter in this case, but it doesn't hurt to start the practice in places where we want just the data.

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts [hunk 2]
```diff
@@ -56,4 +63,22 @@ describe('pipeline-utils ', () => {
...
+    expect(duration).not.toBeNull();
+    expect(duration).toBe('-');
+  });
+
+  it('expect duration to be a time formatted string for PipelineRun with start and end Time', () => {
+    const duration = pipelineRunDuration(mockRunDurationTest[2]);
+    expect(duration).not.toBeNull();
+    expect(duration).not.toBe('-');
+  });
 });
```
Comment (andrewballantyne) [inline frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: Please add success path tests as well. If the tests break at a point in the future, it is nice to know if the whole thing just failed or if just a null-path failed.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx [hunk 3]
```diff
@@ -45,7 +48,7 @@ const PipelineRunRow: React.FC<PipelineRunRowProps> = ({ obj, index, key, style
       <TableData className={tableColumnClasses[3]}>
         <LinkedPipelineRunTaskStatus pipelineRun={obj} />
       </TableData>
-      <TableData className={tableColumnClasses[4]}>-</TableData>
+      <TableData className={tableColumnClasses[4]}>{pipelineRunDuration(obj)}</TableData>
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: @abhinandan13jan Is there any reason for not using [fromNow](https://github.com/openshift/console/blob/master/frontend/public/components/utils/datetime.ts#L4) function?

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: @vikram-raj  If the Pipeline is currently running, `fromNow` would be applicable.
But for Pipelines which have `completed/failed` at a past timestamp. the duration must be the `end-start` time and not `fromNow`

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts (hunk 2 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: This is a success test. This checks for the value to be not `-`. i.e some time string.
`expect(duration).not.toBe('-');`
I am updating this to the exact time-string now.

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts (hunk 1 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: Changing this to use PipelineRunStatus

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx (hunk 3 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: @abhinandan13jan I did not notice that we are using duration for the pipeline.

File: frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts (hunk 2 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: Apparently I am blind... did not see `not`. The test is pretty open-ended, and doesn't cover the values of success. Perhaps we should still write tests to test known success cases.

Comment (andrewballantyne) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: Apparently I am double blind lol - you said `I am updating this to the exact time-string now`. Gosh, that's what I get for starting my day after a long weekend out coffee ;)

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/__tests__/pipeline-utils.spec.ts]: Haha, never mind :)

**Comment by andrewballantyne** [comment]
Please squash (and rebase as needed).

**Comment by abhinandan13jan** [comment]
P.S: unable to squash

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 4]
```diff
@@ -307,3 +307,21 @@ export const getPipelineRunParams = (pipelineParams: PipelineParam[]): PipelineR
...
+  // Duration cannot be computed if start time is missing or a completed/failed pipeline has no end time
+  if (!startTime || (!completionTime && pipelineRunStatus(run) !== 'Running')) {
+    return '-';
+  }
+  const start = new Date(startTime).getTime();
+
+  // For running pipelines duration must be current time - starttime
+  const duration = completionTime
+    ? new Date(completionTime).getTime() - start
+    : new Date().getTime() - start;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts]: This can result in incorrect values due to different time zones.

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: Missed that. Will update

Comment (christianvogt) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: Let me know. This was just a guess.

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: updated this, yes for the current time the browser will pick the local time indeed. Thanks for pointing this out

File: frontend/packages/dev-console/src/utils/pipeline-utils.ts [hunk 5]
```diff
@@ -307,3 +307,24 @@ export const getPipelineRunParams = (pipelineParams: PipelineParam[]): PipelineR
...
+export const pipelineRunDuration = (run: PipelineRun): string => {
+  const startTime = _.get(run, ['status', 'startTime'], null);
+  const completionTime = _.get(run, ['status', 'completionTime'], null);
+
+  // Duration cannot be computed if start time is missing or a completed/failed pipeline has no end time
+  if (!startTime || (!completionTime && pipelineRunStatus(run) !== 'Running')) {
+    return '-';
+  }
+  // Start Time is converted to UTC for correct comparison
+  const start = convertUTC(startTime).getTime();
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/pipeline-utils.ts]: `getTime` is always in UTC.
This current conversion is odd and doesn't do anything different than just using `getTime`.
The question is whether the timestamps returned are in UTC. If yes, then the original implementation you had would work.

Comment (christianvogt) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: I shouldn't have said anything earlier and looked at the unit tests timestamp value which shows the pipeline timestamps are in UTC already. I just confused you.

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-utils.ts]: reverted to original PR

**Pull request merged**